File: apps/uploader/src/uploader/domain/models/sample.py
from __future__ import annotations

from dataclasses import dataclass

from uploader.domain.models.sequencing import SequencingData
from uploader.domain.models.wsi import WsiData


@dataclass(frozen=True)
class Material:
    material_identifier: str | None = None
    collected_from_person: str | None = None
    belongs_to_diagnosis: list[str] | None = None
    sampling_timestamp: str | None = None
    registration_timestamp: str | None = None
    sampling_protocol: str | None = None
    sampling_protocol_deviation: str | None = None
    reason_for_sampling_protocol_deviation: str | None = None
    biospecimen_type: str | None = None
    anatomical_source: str | None = None
    pathological_state: str | None = None
    storage_conditions: str | None = None
    expiration_date: str | None = None
    percentage_tumor_cells: float | None = None
    physical_location: str | None = None
    analyses_performed: list[str] | None = None
    derived_from: str | None = None


@dataclass(frozen=True)
class Sample:
    sample_id: str
    predictive_number: str | None
    bioptic_number: str | None
    payload: dict
    material: Material | None = None
    sequencing: SequencingData | None = None
    wsi: WsiData | None = None

File: src/Uploader/Uploader.Domain/Sample.cs
using System.Text.Json.Nodes;
using Uploader.Domain.Common;

namespace Uploader.Domain;

/// <summary>FAIR Genomes <c>Material</c> value object.</summary>
public sealed record Material : ValueObject
{
    public string? MaterialIdentifier { get; init; }
    public string? CollectedFromPerson { get; init; }
    public IReadOnlyList<string>? BelongsToDiagnosis { get; init; }
    public string? SamplingTimestamp { get; init; }
    public string? RegistrationTimestamp { get; init; }
    public string? SamplingProtocol { get; init; }
    public string? SamplingProtocolDeviation { get; init; }
    public string? ReasonForSamplingProtocolDeviation { get; init; }
    public string? BiospecimenType { get; init; }
    public string? AnatomicalSource { get; init; }
    public string? PathologicalState { get; init; }
    public string? StorageConditions { get; init; }
    public string? ExpirationDate { get; init; }
    public double? PercentageTumorCells { get; init; }
    public string? PhysicalLocation { get; init; }
    public IReadOnlyList<string>? AnalysesPerformed { get; init; }
    public string? DerivedFrom { get; init; }
}

/// <summary>An archived biobank sample plus its derived sequencing/WSI analyses.</summary>
public sealed record Sample : Entity
{
    public required string SampleId { get; init; }
    public string? PredictiveNumber { get; init; }
    public string? BiopticNumber { get; init; }
    public JsonObject? Payload { get; init; }
    public Material? Material { get; init; }
    public IReadOnlyList<SequencingEntry>? Sequencing { get; init; }
    public WsiData? Wsi { get; init; }
}
